cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/ReadThroughBookCache.java

```diff
@@ -0,0 +1,130 @@
+package com.exactpro.cradle.cassandra;
+
+import com.datastax.oss.driver.api.core.cql.BoundStatementBuilder;
+import com.exactpro.cradle.*;
+import com.exactpro.cradle.cassandra.dao.CradleOperators;
+import com.exactpro.cradle.cassandra.dao.books.BookEntity;
+import com.exactpro.cradle.cassandra.dao.books.PageEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+public class ReadThroughBookCache implements BookCache {
+    private static final Logger logger = LoggerFactory.getLogger(ReadThroughBookCache.class);
+
+    private final CradleOperators ops;
+    private final Map<BookId, BookInfo> books;
+    private final Function<BoundStatementBuilder, BoundStatementBuilder> readAttrs;
+    private final String schemaVersion;
+
+    public ReadThroughBookCache(CradleOperators ops,
+                                Function<BoundStatementBuilder, BoundStatementBuilder> readAttrs,
+                                String schemaVersion) {
+        this.ops = ops;
+        this.books = new ConcurrentHashMap<>();
+        this.readAttrs = readAttrs;
+        this.schemaVersion = schemaVersion;
+    }
+
+    public BookInfo getBook (BookId bookId) {
+        if (!books.containsKey(bookId)) {
+            logger.info("Could not find book named {} in cache, tyring getting from db", bookId.getName());
+            try {
+                books.put(bookId, loadBook(bookId));
+            } catch (Exception e) {
+                logger.warn("Could not find book named {} in database", bookId.getName());
+            }
+        }
+
+        return books.get(bookId);
+    }
+
+    public Collection<PageInfo> loadPageInfo(BookId bookId) throws IOException
+    {
+        Collection<PageInfo> result = new ArrayList<>();
+        try
+        {
+            for (PageEntity pageEntity : ops.getOperators(bookId, readAttrs).getPageOperator().getAll(bookId.getName(), readAttrs))
+            {
+                if (pageEntity.getRemoved() == null)
+                    result.add(pageEntity.toPageInfo());
+            }
+        }
+        catch (Exception e)
+        {
+            throw new IOException("Error while loading pages of book '"+bookId+"'", e);
+        }
+        return result;
+    }
+
+    public BookInfo loadBook (BookId bookId) throws IOException {
+        try {
+            BookEntity bookEntity = ops.getCradleBookOperator().get(bookId.getName(), readAttrs);
+
+            if (!bookEntity.getSchemaVersion().equals(schemaVersion)) {
+                throw new IOException(String.format("Unsupported schema version for the book \"%s\". Expected: %s, found: %s. Skipping",
```

Comment:
This is actually not an `IOException`. And you are not skipping it, you throw an exception.

Reply:
Resolved, [8d758ea](https://github.com/th2-net/cradleapi/pull/85/commits/8d758ea710b501ede06ef3cf1cb5bb4a1699c435)